### PR TITLE
Jesse/Issue 4

### PR DESCRIPTION
**Pull request by vc-ca**
Sockets now run a JWT verify check, to see if client has a valid token. 
The decoded token response is then set as the socket.id.
New messages now emit to a specific id, rather than broadcasting to everyone.

My goal with these two items was to:
1. Allow for there to be some auth check on the sockets.
2. Make recipient ID's available, so if the app was built out to include multiple concurrent users, you'd be able to message directly to a single user, rather than all users.

### CODE REVIEW

File: client/src/store/utils/reducerFunctions.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 export const addMessageToStore = (state, payload) => {
   const { message, sender } = payload;
+  console.log("message text", message.text);
```
Comment (RayK3) [inline client/src/store/utils/reducerFunctions.js:3]: Looks like you forgot to remove a little `console.log` here :smile:

File: server/bin/www [hunk 2]
```diff
@@ -31,17 +31,32 @@ const server = http.createServer(app);
 
 const io = require("socket.io")(server);
 
-io.on("connection", (socket) => {
+io.use((socket, next) => {
+  if (socket.handshake.headers.cookie) {
+    let token = socket.handshake.headers.cookie;
+    //remove token= off the front of the string
+    let slicedToken = token.slice(6);
```
Comment (RayK3) [inline server/bin/www:38]: It seems like both of these `let` statements can be `const` 👍

File: server/bin/www [hunk 3]
```diff
@@ -31,17 +31,32 @@ const server = http.createServer(app);
...
+    jwt.verify(slicedToken, process.env.SESSION_SECRET, (err, decoded) => {
+      if (err) return next(new Error("Authentication error"));
+      socket.decoded = decoded;
```
Comment (RayK3) [inline server/bin/www:41]: I don't think you need to store the entire token onto the socket object. Also where are you using this `socket.id` after you save it?

File: server/bin/www [hunk 4]
```diff
@@ -31,17 +31,32 @@ const server = http.createServer(app);
...
+  }
+}).on("connection", (socket) => {
   socket.on("go-online", (id) => {
     if (!onlineUsers.includes(id)) {
       onlineUsers.push(id);
     }
-    // send the user who just went online to everyone else who is already online
     socket.broadcast.emit("add-online-user", id);
+    console.log("socket decoded", socket.decoded);
+    console.log("id", socket.id);
```
Comment (RayK3) [inline server/bin/www:55]: These ones as well 🙂

**Review by RayK3 — COMMENTED** [review]
Seems like you are doing socket auth properly, and are also only emitting messages to conversation members, so nice job :100:

There were a couple of `console.log`s left in that can be removed, and I also had a question about using the `socket.id` 😄

File: server/bin/www (hunk 3 above)
Comment (vc-ca) [reply server/bin/www:41]: Ahh this is true. removed socket.decoded. 

Here was my thinking. By resaving the socket.id from the randomly generated string to the user id from the chat app:
```
jwt.verify(slicedToken, process.env.SESSION_SECRET, (err, decoded) => {
      if (err) return next(new Error("Authentication error"));
      socket.id = decoded.id;
      next();
    });
```
  Later on, when sending a direct message, I can send it to the numeric data.recipientId, which will match up with the socket.id issued on login.
   
```
    socket.on("new-message", (data) => {
    socket.broadcast.to(data.recipientId).emit("new-message", {
      message: data.message,
      sender: data.sender,
      recipientId: data.recipientId,
    });
  });
```

**Review by bonnieli — APPROVED** [review]

**Pull request merged**